source/static_thread_pool.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <unifex/static_thread_pool.hpp>
+
+namespace unifex
+{
+  static_thread_pool::static_thread_pool()
+    : static_thread_pool(std::thread::hardware_concurrency()) {}
+
+  static_thread_pool::static_thread_pool(std::uint32_t threadCount)
+    : threadCount_(threadCount)
+    , threadStates_(threadCount)
+    , nextThread_(0) {
+    assert(threadCount > 0);
+
+    threads_.reserve(threadCount);
+    try {
+      for (std::uint32_t i = 0; i < threadCount; ++i) {
+        threads_.emplace_back([this, i] { run(i); });
+      }
+    } catch (...) {
+      request_stop();
+      join();
+      throw;
+    }
+  }
+
+  static_thread_pool::~static_thread_pool() {
+    request_stop();
+    join();
+  }
+
+  void static_thread_pool::request_stop() noexcept {
+    for (auto& state : threadStates_) {
+      state.request_stop();
+    }
+  }
+
+  void static_thread_pool::run(std::uint32_t index) noexcept {
+    while (true) {
+      task_base* task = nullptr;
+      for (std::uint32_t i = 0; i < threadCount_; ++i) {
+        auto queueIndex = (index + i) < threadCount_
+            ? (index + i)
+            : (index + i - threadCount_);
+        auto& state = threadStates_[queueIndex];
+        task = state.try_pop();
+        if (task != nullptr) {
+          break;
+        }
+      }
+
+      if (task == nullptr) {
+        task = threadStates_[index].pop();
+        if (task == nullptr) {
+          // request_stop() was called.
+          return;
+        }
+      }
+
+      task->execute(task);
+    }
+  }
+
+  void static_thread_pool::join() noexcept {
+    for (auto& t : threads_) {
+      t.join();
+    }
```

Comment:
If you used `jthread` this loop would be unnecessary. 🤣

Reply:
Indeed :)
I've heard that libstdc++ will have jthread/stop_token soon.
I haven't heard about libc++/msvc yet.